README.md
# pyotp2289

*pyotp2289* is a pure Python 3 implementation of "A One-Time Password System" -
RFC-2289.

It requires no additional libraries.

## General

The main reason for writing this library was the need to login into my
FreeBSD servers using [opiepasswd(1)](https://www.freebsd.org/cgi/man.cgi?query=opiepasswd&sektion=1&manpath=FreeBSD)
as [described in the FreeBSD Handbook](https://docs.freebsd.org/en/books/handbook/security/#one-time-passwords).

I decided to license the library under the
[Simplified BSD License / 2-clause BSD license](https://github.com/blackm0re/pyotp2289/blob/master/LICENSE) and not under the
(L)GPL-3 as I usually do.

I hope that somebody will find it useful.

## Installation

### pip (pypi)

   ```bash
   pip install pyotp2289
   ```

### FreeBSD

*pyotp2289* is included in the official ports-tree.

   ```bash
   pkg install security/py-pyotp2289
   ```

– or: 

   ```bash
   cd /usr/ports/security/py-pyotp2289
   make install clean
   ```

### Gentoo

   ```bash
   # add sgs' custom repository using app-eselect/eselect-repository
   eselect repository add sgs

   # ... or using app-portage/layman (obsolete)
   layman -a sgs

   emerge dev-python/pyotp2289
   ```

## Overview of RFC-2289

RFC-2289 describes a one-time password authentication system (OTP):

"The system provides authentication for system access (login) and other
applications requiring authentication that is secure against passive attacks
based on replaying captured reusable passwords. OTP evolved from the S/KEY
(S/KEY is a trademark of Bellcore) One-Time Password System that was released
by Bellcore."

"One form of attack on networked computing systems is eavesdropping on
network connections to obtain authentication information such as the
login IDs and passwords of legitimate users. Once this information is
captured, it can be used at a later time to gain access to the
system. One-time password systems are designed to counter this type
of attack, called a 'replay attack'."

The authentication system described in RFC-2289 "uses a secret
pass-phrase to generate a sequence of one-time (single use)
passwords.  With this system, the user's secret pass-phrase never
needs to cross the network at any time such as during authentication
or during pass-phrase changes. Thus, it is not vulnerable to replay
attacks.  Added security is provided by the property that no secret
information need be stored on any system, including the server being
protected."

"There are two entities in the operation of the OTP one-time password
system. The **generator** must produce the appropriate one-time password
from the user's secret pass-phrase and from information provided in
the **challenge** from the **server**. The server must send a challenge that
includes the appropriate generation parameters to the generator, must
verify the one-time password received, must store the last valid
one-time password it received, and must store the corresponding one-
time password sequence number. The server must also facilitate the
changing of the user's secret pass-phrase in a secure manner."

"The OTP system generator passes the user's secret pass-phrase, along
with a seed received from the server as part of the challenge,
through multiple iterations of a secure hash function to produce a
one-time password. After each successful authentication, the number
of secure hash function iterations is reduced by one.  Thus, a unique
sequence of passwords is generated.  The server verifies the one-time
password received from the generator by computing the secure hash
function once and comparing the result with the previously accepted
one-time password."


## Examples

We define the two entities: *client* and *server*. The entire application of
RFC-2289 consists of interactions between them.

   ```python
   import getpass  # client only

   import otp2289  # client and server

   # the server starts by picking:
   # - algorithm (MD5 or SHA1) to use
   # - seed - 1 to 16 alphanumeric characters. The seed must never be reused.
   # - initial step - number (int) that will be decremented for each OTP.
   # In FreeBSD, the following default values are used:
   # - MD5
   # - the first two letters of the hostname + 5 random digits for seed
   # - initial step: 500

   # the client receives those values, chooses a strong password and creates
   # initialization digest (hash). The password 'This is a test.' will give you
   # the same results as in the following example.
   passwd_bytes = getpass.getpass().encode()  # Fetch the password as bytes
   generator = otp2289.generator.OTPGenerator(passwd_bytes,
                                              'TesT',
                                              otp2289.OTP_ALGO_MD5)
   digest = generator.generate_otp_hexdigest(500)
   # digest is now: 0x2b8d82b6ac14346c
   # the client sends it to the server

   # the server creates the first state. Note that step is decremented by 1:
   state = otp2289.server.OTPState(digest, 499, 'TesT', otp2289.OTP_ALGO_MD5)
   # the state can be stored in a OTPStore container:
   store = otp2289.server.OTPStore()
   # key can be any str that can be used to reference the state (f.i username)
   store.add_state('myusername', state)  # where key can be any str that can be
   # OTPStore is provided only for convenience as it is not part of RFC-2289.
   # The server can store states any way it wants. A normal dict is also fine.
   # Once the initial state is set on the server, the client can authenticate.

   # Upon authentication request (f.i. login), the server issues a challenge
   # based on the state:
   challenge = state.challenge_string  # challenge is now 'otp-md5 499 TesT '

   # the client can now respond by using (or recreating) the same generator
   # created earlier. RFC-2289 defines two types of responses:
   # - hex (like '0x2b8d82b6ac14346c') - more suited for automation
   # - tokens consisting of 6 short words - better when responding manually
   hex_response = generator.generate_otp_hexdigest(499)  # '0x6323f96296a2526b'
   token_response = generator.generate_otp_words(499)
   # token_response is now: 'CANT JAW BITS NU LO PUP'
   # a possible shortcut may be to use the challenge-string directly:
   hex_response = generator.generate_otp_hexdigest_from_challenge(challenge)
   token_response = generator.generate_otp_words_from_challenge(challenge)
   # ... giving the same results.

   # once the response is received, the server validates it by yet again using
   # the current state:
   result = state.response_validates(hex_response)
   # or
   result = state.response_validates(token_response)
   # result should be True if the response matches the state, False if not
   # in case of invalid response or response checksum doesn't match, a
   # otp2289.server.OTPInvalidResponse exception is raised.

   # once the state has successfully validated the corresponding response,
   # the state **must never be used again** and a state corresponding to the
   # "next" (498) step created.
   state = state.get_next_state()

   # the next authentication attempt...
   challenge = state.challenge_string  # challenge is now 'otp-md5 498 TesT '
   # ... and on the client side...
   hex_response = generator.generate_otp_hexdigest_from_challenge(challenge)
   # etc. etc...
   ```

Please visit the
[API documentation](http://gnulover.simeonov.no/docs/api/pyotp2289/latest/) for
a complete reference.

If you don't care about developing applications in Python and only care about
generating one-time passwords (tokens / hex digests) and authenticating with
existing solutions (f.i. FreeBSD servers), *pyotp2289* comes with a simple CLI:

   ```bash
   python -m otp2289 --generate-otp-response -f token -i 498 -s TesT
   ```

... will prompt for password and generate a 6 words (token) response.

   ```bash
   python -m otp2289 --generate-otp-range -f token -i 498 -s TesT
   ```

... will prompt for password and generate a range of 4 one-time passwords
starting from (and including) 498.


## Support and contributing

*pyotp2289* is hosted on GitHub: https://github.com/blackm0re/pyotp2289


## Author

Simeon Simeonov - sgs @ LiberaChat


## [License](https://github.com/blackm0re/pyotp2289/blob/master/LICENSE)

Copyright (c) 2020-2023 Simeon Simeonov
All rights reserved.

[Licensed](https://github.com/blackm0re/pyotp2289/blob/master/LICENSE) under the BSD 2-clause.
SPDX-License-Identifier: BSD-2-Clause-FreeBSD


CHANGELOG.md
# Changelog

## [1.2.1](https://github.com/blackm0re/pyotp2289/tree/1.2.1) (2023-01-08)

[Full Changelog](https://github.com/blackm0re/pyotp2289/compare/1.2.0...1.2.1)

**Changes:**

- Include the unit tests in the sdist package


## [1.2.0](https://github.com/blackm0re/pyotp2289/tree/1.2.0) (2023-01-05)

[Full Changelog](https://github.com/blackm0re/pyotp2289/compare/1.1.1...1.2.0)

**Changes:**

- Add the `OTPState.ot_hex` property


## [1.1.1](https://github.com/blackm0re/pyotp2289/tree/1.1.1) (2022-04-02)

[Full Changelog](https://github.com/blackm0re/pyotp2289/compare/1.1.0...1.1.1)

**Changes:**

- Reintroduce setup.py for better campatibility


## [1.1.0](https://github.com/blackm0re/pyotp2289/tree/1.1.0) (2022-03-29)

[Full Changelog](https://github.com/blackm0re/pyotp2289/compare/1.0.0...1.1.0)

**Changes:**

- (CLI) a password can now be set using the *OTP2289_PASSWORD* env. variable

- (CLI) a new parameter *-P* / *--force-password-prompt* can be used in order to force password prompt

- improved tests for CLI

- use *setuptools* instead of the deprecated *distutils*


# [1.0.0](https://github.com/blackm0re/pyotp2289/tree/1.0.0) (2020-04-07)

**Changes:**

- Initial release


src/otp2289/__main__.py
# -*- coding: utf-8 -*-
# SPDX-License-Identifier: BSD-2-Clause-FreeBSD
#
# Copyright (c) 2020-2023 Simeon Simeonov
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
# NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""
CLI entry point for the otp2289 package

Examples:
python -m otp2289 --initiate-new-sequence -s TesT

python -m otp2289 --generate-otp-response -c "otp-md5 499 TesT " -f token
python -m otp2289 --generate-otp-response -s TesT -i 499 -f token
"""
import argparse
import errno
import getpass
import io
import os
import secrets
import string
import sys

import otp2289


def eprint(*arg, **kwargs):
    """stdderr print wrapper"""
    print(*arg, file=sys.stderr, flush=True, **kwargs)


def generate_otp_response(args: argparse.Namespace) -> str:
    """
    Generates a response based on the parameters sent from the parser

    :param args: The arguments assigned from argparse
    :type args: argparse.Namespace

    :raises otp2289.OTPChallengeException: If the challenge is invalid

    :raises otp2289.OTPGeneratorException: If generator parameters are wrong

    :return: The response string
    :rtype: str
    """
    generator = otp2289.generator.OTPGenerator(
        args.password.encode(),
        args.seed,
        args.hash_algo,
    )
    if args.challenge_string:
        if args.output_format == 'token':
            return generator.generate_otp_words_from_challenge(
                args.challenge_string
            )
        return generator.generate_otp_hexdigest_from_challenge(
            args.challenge_string
        )
    # regular parameters
    header = ''
    if not args.quiet:
        header = (
            f'Seed: {args.seed}, Step: {args.step}, '
            f'Hash: {args.hash_algo}{os.linesep}'
        )
    if args.output_format == 'token':
        return header + generator.generate_otp_words(args.step)
    return header + generator.generate_otp_hexdigest(args.step)


def generate_otp_range(args: argparse.Namespace) -> str:
    """
    Generates range of responses based on the parameters sent from the parser

    :param args: The arguments assigned from argparse
    :type args: argparse.Namespace

    :raises otp2289.OTPChallengeException: If the challenge is invalid

    :raises otp2289.OTPGeneratorException: If generator parameters are wrong

    :return: The responses string
    :rtype: str
    """
    generator = otp2289.generator.OTPGenerator(
        args.password.encode(),
        args.seed,
        args.hash_algo,
    )
    if args.output_format == 'token':
        method = generator.generate_otp_words
    else:
        method = generator.generate_otp_hexdigest
    # handle most cases explicitly
    if args.range == 1:
        return f'{args.step}: ' + method(args.step)
    if args.range > args.step + 1:
        args.range = args.step + 1
    # any need for quiet?
    header = ''
    if not args.quiet:
        header = (
            f'Seed: {args.seed}, Step: {args.step}, '
            f'Hash: {args.hash_algo}, Range: {args.range}{os.linesep}'
        )
    return header + os.linesep.join(
        [
            f'{step}: ' + method(step)
            for step in range(
                args.step,
                args.step - args.range,
                -1,
            )
        ]
    )


def get_password(args: argparse.Namespace) -> str:
    """
    Extract the provided password using the defined argparse arguments

    :param args: The arguments assigned from argparse
    :type args: argparse.Namespace

    :raises KeyboardInterrupt: If the password prompt is interrupted

    :return: The extrated password string
    :rtype: str
    """
    if args.force_password_prompt:
        while True:
            password = getpass.getpass()
            if not args.initiate_new_sequence or password == getpass.getpass(
                'Repeat password: '
            ):
                break
            eprint('The passwords do not match')
        return password

    if not args.password:
        password = os.environ.get('OTP2289_PASSWORD')
        if password is not None:
            return password
        while True:
            password = getpass.getpass()
            if not args.initiate_new_sequence or password == getpass.getpass(
                'Repeat password: '
            ):
                break
            eprint('The passwords do not match')
        return password

    if os.path.isfile(args.password):
        with io.open(args.password, 'r', encoding='utf-8') as fp:
            return fp.readline().strip()

    return args.password


def get_rnd_seed() -> str:
    """
    Returns a random seed in the format:

    2 random letters (capitalize()) + 5 random digits
    """
    rnd = secrets.SystemRandom()
    return ''.join(
        rnd.choices(string.ascii_lowercase, k=2)
    ).capitalize() + ''.join(rnd.choices(string.digits, k=5))


def initiate_new_sequence(args: argparse.Namespace) -> str:
    """
    Generates a new sequence based on the parameters sent from the parser.

    :param args: The arguments assigned from argparse
    :type args: argparse.Namespace

    :raises otp2289.OTPChallengeException: If the challenge is invalid

    :raises otp2289.OTPGeneratorException: If generator parameters are wrong

    :return: The response string
    :rtype: str
    """
    if not args.seed:
        args.seed = get_rnd_seed()
    header = ''
    if not args.quiet:
        header = (
            f'Seed: {args.seed}, Step: {args.step}, '
            f'Hash: {args.hash_algo}{os.linesep}'
        )
    generator = otp2289.generator.OTPGenerator(
        args.password.encode(),
        args.seed,
        args.hash_algo,
    )
    if args.challenge_string:
        return header + generator.generate_otp_hexdigest_from_challenge(
            args.challenge_string
        )
    return header + generator.generate_otp_hexdigest(args.step)


def main(args=None):
    """the main entry point"""
    parser = argparse.ArgumentParser(
        prog=__package__,
        epilog=(
            f'%(prog)s {otp2289.__version__} by Simeon Simeonov '
            '(sgs @ LiberaChat)'
        ),
        description='The following options are available',
    )
    group = parser.add_mutually_exclusive_group(required=True)
    password_group = parser.add_mutually_exclusive_group()
    group.add_argument(
        '--generate-otp-range',
        action='store_true',
        dest='generate_otp_range',
        default=False,
        help='Generates a range of OTP responses',
    )
    group.add_argument(
        '--generate-otp-response',
        action='store_true',
        dest='generate_otp_response',
        default=False,
        help='Generates a new OTP response',
    )
    group.add_argument(
        '--initiate-new-sequence',
        action='store_true',
        dest='initiate_new_sequence',
        default=False,
        help=(
            'Initiates a new OTP sequence. Essentially the same as '
            '--generate-otp-response only it prompts twice for password '
            'and always outputs hex (ignores -f).'
        ),
    )
    password_group.add_argument(
        '-P',
        '--force-password-prompt',
        dest='force_password_prompt',
        action='store_true',
        help=(
            'Force password prompt even if the env. variable '
            '"OTP2289_PASSWORD" is set'
        ),
    )
    password_group.add_argument(
        '-p',
        '--password',
        metavar='<PASSWORD[FILE]>',
        type=str,
        dest='password',
        default='',
        help=(
            'The password or path to password file '
            '(default & recommended: prompt for passwd)'
        ),
    )
    parser.add_argument(
        '-a',
        '--hash-algorithm',
        metavar='<md5 | sha1>',
        type=str,
        dest='hash_algo',
        default='md5',
        help='The hash algorithm to use. Possible values: md5 (default), sha1',
    )
    parser.add_argument(
        '-c',
        '--challenge-string',
        metavar='<challenge string>',
        type=str,
        dest='challenge_string',
        default='',
        help='Use challenge string when generating response',
    )
    parser.add_argument(
        '-f',
        '--output-format',
        metavar='<hex | token>',
        type=str,
        dest='output_format',
        default='hex',
        help='The output format to use. Possible values: hex (default), token',
    )
    parser.add_argument(
        '-i',
        '--step',
        metavar='<step>',
        type=int,
        dest='step',
        default=500,
        help='The step. Default for initiating a new sequence is: 500',
    )
    parser.add_argument(
        '-q',
        '--quiet',
        action='store_true',
        dest='quiet',
        default=False,
        help='Dot not show headers. Only hex / tokens',
    )
    parser.add_argument(
        '-r',
        '--range',
        metavar='<range>',
        type=int,
        dest='range',
        default=1,
        help='Amount of consecutive OTP hex/tokens to generate. default: 1',
    )
    parser.add_argument(
        '-s',
        '--seed',
        metavar='[seed]',
        type=str,
        dest='seed',
        default='',
        help=(
            'The seed to use (1 to 16 alphanumeric characters) '
            '(default & recommended: random seed)'
        ),
    )
    parser.add_argument(
        '-v',
        '--version',
        action='version',
        version=f'%(prog)s {otp2289.__version__}',
        help='display program-version and exit',
    )
    args = parser.parse_args(args)
    # handle the password before everything else
    try:
        args.password = get_password(args)
    except KeyboardInterrupt:
        eprint(os.linesep + 'Prompt terminated')
        sys.exit(errno.EACCES)
    except Exception as exp:
        eprint(f'Unable to fetch password: {exp}')
        sys.exit(1)
    try:
        if args.initiate_new_sequence:
            print(initiate_new_sequence(args))
        if args.generate_otp_range:
            print(generate_otp_range(args))
        if args.generate_otp_response:
            print(generate_otp_response(args))
        sys.exit(0)
    except otp2289.generator.OTPGeneratorException as exp:
        eprint(f'GeneratorException: {exp}')
    except otp2289.generator.OTPChallengeException as exp:
        eprint(f'ChallengeException: {exp}')
    except Exception as exp:
        eprint(f'Unknown error: {exp}')
    sys.exit(1)


if __name__ == '__main__':
    main()
